spec/setfrozentag_action_spec.rb

```diff
@@ -0,0 +1,136 @@
+    end
+
+    context 'with github_token' do
+      it 'properly passes the environment variable `GITHUB_TOKEN` all the way to Octokit::Client' do
+        ENV['GITHUB_TOKEN'] = test_token
+        expect(Octokit::Client).to receive(:new).with(access_token: test_token)
+        run_described_fastlane_action(mock_params)
+      end
+
+      it 'properly passes the parameter `:github_token` all the way to Octokit::Client' do
+        mock_params[:github_token] = test_token
+        expect(Octokit::Client).to receive(:new).with(access_token: test_token)
+        run_described_fastlane_action(mock_params)
+      end
+
+      it 'prioritizes `:github_token` parameter over `GITHUB_TOKEN` enviroment variable if both are present' do
+        ENV['GITHUB_TOKEN'] = 'Test-EnvGithubToken-1234'
+        mock_params[:github_token] = test_token
+        expect(Octokit::Client).to receive(:new).with(access_token: test_token)
+        run_described_fastlane_action(mock_params)
+      end
+
+      it 'prints an error if no `GITHUB_TOKEN` environment variable nor parameter `:github_token` is present' do
+        expect { run_described_fastlane_action(mock_params) }.to raise_error(FastlaneCore::Interface::FastlaneError)
+      end
+    end
+
+    context 'with default parameters' do
+      let(:glithub_helper) do
+        instance_double(
+          Fastlane::Helper::GithubHelper,
+          get_milestone: test_milestone,
+          create_milestone: nil
+        )
+      end
+
+      before do
+        mock_params[:github_token] = test_token
+        allow(Fastlane::Helper::GithubHelper).to receive(:new).and_return(glithub_helper)
+      end
+
+      it 'froze the milestone when the parameter `:freeze` is not present' do
+        default_freeze_milestone = { title: '10.1 ❄️' }
+        expect(glithub_helper).to receive(:update_milestone).with(
+          repository: anything,
+          number: anything,
+          options: default_freeze_milestone
+        )
+        run_described_fastlane_action(mock_params)
+      end
+    end
+  end
+
+  describe 'get_milestone' do
+    let(:test_repository) { 'test-repository' }
+    let(:test_milestone) { { title: '10.1', number: 1234 } }
+    let(:mock_params) do
+      {
+        repository: test_repository,
+        milestone: test_milestone[:title],
+        github_token: 'Test-GithubToken-1234'
+      }
+    end
+    let(:client) do
+      instance_double(
+        Octokit::Client,
+        list_milestones: [test_milestone],
+        update_milestone: nil,
+        user: instance_double('User', name: 'test'),
+        'auto_paginate=': nil
+      )
+    end
+
+    before do
+      allow(Octokit::Client).to receive(:new).and_return(client)
+    end
+
+    it 'properly passes parameters all the way down to the Octokit::Client' do
+      expect(client).to receive(:list_milestones).with(test_repository)
+      run_described_fastlane_action(mock_params)
+    end
+  end
+
+  describe 'update_milestone' do
```

Comment:
Self-Review: This is also a point where I have doubts about the pattern used across the app, I leave here the name of the method that Oktokit should call to pass the correct data. I was wondering at this moment if I should instead have done as:

```ruby
describe 'run' do #or executeAction (naming is hard here)
   it 'update_milestone properly passes parameters all the way down to the Octokit::Client'
     ...
   end
end

```

Reply:
Yeah as commented above, I'm having a hard time following the logic of how you split and organized the tests tbh.

Here since the goal is to unit test the `SetfrozentagAction`, why do we even have a `describe 'update_milestone'` (which relates to an API method here, while `describe` is usually more to group together multiple unit tests that test one common thing but stressing it in multiple ways)?

What we want to test in `setfrozentag_action_spec` is that, given various different inputs passed to the action (different parameters passed to `run_described_fastlane_action(…)`), the final outcome is what's expected — i.e. the final method call on the `Octokit::Client` that updates the milestone with the frozen emoji updates the right milestone, and with the right expected title. And, that if we pass invalid parameters or encounter an error or corner case (e.g. we're asking the action to `setfrozentag(milestone: '10.1', …)` but no milestone with title `10.1` exists), it handles it properly/nicely instead of doing something unexpected (like modifying another milestone or crashing badly).